test/es-module/test-esm-blob-urls.js

```diff
@@ -0,0 +1,77 @@
+const { test } = require('node:test');
+
+async function createURL(mime, body) {
+  const blob = new Blob([body], { type: mime });
+  return URL.createObjectURL(blob);
+}
+
+function createBase64URL(mime, body) {
+  return `data:${mime};base64,${Buffer.from(body).toString('base64')}`;
+}
+
+test('importing JavaScript from a Blob: URL should work', async (t) => {
+  const url = await createURL('text/javascript', 'export default { "key": "value" }');
+  const { default: expected } = await import(url);
+  t.assert.deepStrictEqual(expected, { key: 'value' });
+});
+
+test('import.meta.url should be equal to the blob URL', async (t) => {
+  const url = await createURL('text/javascript', 'export default import.meta.url;');
+  const { default: expected } = await import(url);
+  t.assert.deepStrictEqual(expected, url);
+});
+
+
+test('importing JSON from a Blob: URL should work', async (t) => {
+  const url = await createURL('application/json', '{ "key": "value" }');
+  const { default: expected } = await import(url, { with: { type: 'json' } });
+  t.assert.deepStrictEqual(expected, { key: 'value' });
+});
+
+test('importing a module from within a blob: URL should work', async (t) => {
+  await t.test('file: URLs', async (t) => {
+    const url = await createURL('text/javascript', `import { message } from ${JSON.stringify(fixtures.fileURL('es-modules/message.mjs'))}; export default message;`);
+    const { default: expected } = await import(url);
+    t.assert.strictEqual(expected, 'A message');
+  });
+
+  await t.test('data: URLs', async (t) => {
+    const dataURL = createBase64URL('text/javascript', 'export const message = "Base64 Imported";');
+    const url = await createURL('text/javascript', `import { message } from "${dataURL}"; export default message;`);
+    const { default: expected } = await import(url);
+    t.assert.strictEqual(expected, 'Base64 Imported');
+  });
+
+  await t.test('blob: URLs', async (t) => {
+    const url = await createURL('text/javascript',
+                                `const blob = new Blob(["export default 'Blob Imported'"], { type: 'application/javascript' });` +
+      'const url = URL.createObjectURL(blob);' +
+      'export default await import(url);'
+    );
+    const { default: { default: expected } } = await import(url);
+    t.assert.strictEqual(expected, 'Blob Imported');
+  });
+});
+
+test('importing a disallowed mimetype should fail', async (t) => {
+  const url = await createURL('invalid/mime', '');
+  await t.assert.rejects(import(url), /ERR_UNKNOWN_MODULE_FORMAT/);
+});
+
+test('importing a blob from a hook should work', async (t) => {
+  const spawned = await common.spawnPromisified(process.execPath, [
+    '--import',
+    'data:text/javascript,import { register } from "node:module"; import { pathToFileURL } from "node:url";' +
+    `register(${JSON.stringify(fixtures.fileURL('es-module-loaders/hooks-blobify.mjs'))}, pathToFileURL("./"));`,
+    '--import', 'data:load_blob',
+    '--eval=""',
```

Comment:
We don't need the fixtures, what I want to validate is whether it's possible to load a blob: URL generated on the main thread from the loader thread.
```suggestion
    `register("data:text/javascript,");`,
    '--import', 'data:load_blob',
    '--input-type=module',
    '--eval',
    `${createURL}await import(await createURL('text/javascript', 'console.log("Loaded");'));`,
```

Reply:
IIUC blobs are limited to the process that they created from, and so that won't work.

```
import { resolveObjectURL } from 'buffer';
import { fileURLToPath } from 'url';
import { Worker, isMainThread, parentPort, workerData } from 'worker_threads';

if (isMainThread) {
    const value = await URL.createObjectURL(new Blob(['Data']));
    const worker = new Worker(fileURLToPath(import.meta.url), {
        workerData: { value }
    });

    console.log("MAIN", value, resolveObjectURL(value));
} else {
    const { value } = workerData;
    console.log("WORKER", value, resolveObjectURL(value));
}
```

Reply:
It has documented, but if you aren't satisfied, LMK

Reply:
That's of course very dissatisfying if blob URLs are completely broken as soon as a hook is registered, IMO we shouldn't ship something that's going to break like that.

Reply:
If I'm understanding how hooks work (like how I had this setup before), blob URLs will work, they just won't work from the CLI.

Reply:
My suggestion is not to pass them via the CLI, just to have a test that validates blob URLs still work when there are loaders registered

Reply:
After looking into it, this is not possible, so the following note has been added to the docs: "Blob URLs are registered within the current thread. Loaders will not be able to handle resolve Blob's from the main thread."